### PR TITLE
Rework the HoS's Energy Shotgun (Varying energy consumption depending on fire-mode + re-adds a toned down self recharge.)

### PR DESCRIPTION
**Pull request by BramvanZijp**
## About the PR
This PR changes the HoS's Energy shotgun in 2 main ways:
1. The capacity has been set to 1200, and the consumption per shot has been set to 120 for disabler (10 shots), 150 for wide spread (8 shots) and 200 for narrow spread (6 shots)

2. The self recharge feature has been readded, it takes 50 seconds to fully recharge, BUT the weapon does not self recharge if it has been used within the last 10 seconds, meaning you need to fully stop using it for 60 seconds to recharge it. 

## Why / Balance
Alright, here we go, time for me to justify readding the self recharge.
The energy shotgun in an ideal world is a weapon that while not being too overpowered, like what it was when it had the original self recharge, is still a substantial item which is worth to carry for the HoS and gives them a unique ability.
While currently the item is not in that OP state anymore, it is now so weak that most HoS's leave it in favor of their WT550, either stuffing it in their locker or handing it down to the warden or another security officer, which defeats the purpose of it being the HoS's high value steal objective item.

This PR aims to change that by re-adding the self recharge without being as OP as it was.
The fact that you need to stop using the weapon for it to be able to recharge means that it still has substantial downtime, so while it has infinite ammo on paper, in practice you will need to either take cover for a whole minute or switch to another weapon every time your charges run dry if you are in a sustained firefight such as in a nukie round.

Furthermore, narrow spread has always been better then wide as your standard, but with it getting 2 extra shots, it should be more worth using now in close range, and disabler may also see more use with its fire mode now having 10 charges.

These changes were discussed in the #design-discussion channel in the SS14 discord, and the general consensus was that it is worth trying this out to see how it is, and then from there we can better decide if it needs to be tweaked or reverted, since the in-game implications of these changes can only be estimated so much without seeing it in action in real rounds.

## Technical details
I'll start off by saying if the code needs to be cleaned up before this is merged, I get that.
The BatterySelfRechargerComponent now has 3 extra fields, all of which have defaults to avoid messing up other items that use it.

autoRechargePause is a bool which defaults to false, and as such by default this does not change how self recharge works unless set to true.
autoRechargePauseTime, a float defaulted to 0, decides how long the auto recharge should be on cooldown for after a charge is used from the battery.
autoRechargeCooldown is a float defaulted to 0, every time the item uses a charge, it is set to be equal to PauseTime, and if this is more then 0 while autoRechargePause is true, the item does not self recharge, and instead autoRechargeCooldown is reduced by frameTime, meaning it drops 1 every second.

## Media
https://github.com/user-attachments/assets/c65e23b1-4e8f-4505-8982-4eb56dff3f48

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

**Changelog**
:cl: BramvanZijp
- add: The Head of Security's Energy Shotgun has regained its ability to self recharge, but only if the weapon has not been discharged in the last 10 seconds.
- tweak: The Head of Security's Energy Shotgun now has a varying amount of shots depending on the firing mode.
- fix: Energy based weapons with differing energy consumption depending on the fire-mode will now update the ammo counter on the HUD when switching between fire-modes.

### CODE REVIEW

**Comment by DieselMohawk** [comment]
I hope this means it won't get relegated to wardens anymore

**Comment by BramvanZijp** [comment]
> The amount of changes the HoS shotgun is getting

It started OP, it got changed to underpowered, hopefully this PR puts it in a healthy balanced state.

**Comment by ScarKy0** [comment]
> It started OP, it got changed to underpowered, hopefully this PR puts it in a healthy balanced state.

Yeah based on the changes im seeing itll definitely be good but not overwhelmingly good
The shotgun really needs that self recharge

**Comment by PrivateJoe09** [comment]
I really think this is a great idea to not leave the shotgun in the armory/give it to the warden, I have done that every round after the nerf.

**Comment by Plykiya** [comment]
10 seconds still seems like way too little time for the recharge cooldown. This still fully allows an antag to do exactly what they were doing with it before, which is just, kill one or two people, run into maints, and then kill another one or two by the time they exit maints.

**Comment by BramvanZijp** [comment]
> 10 seconds still seems like way too little time for the recharge cooldown. This still fully allows an antag to do exactly what they were doing with it before, which is just, kill one or two people, run into maints, and then kill another one or two by the time they exit maints. They only need to do this two or three times to take out an entire sec force, and it takes way longer to heal people than for it to recharge.

While the numbers can be tweeked, I feel like its enough currently. And if sec let them steal it, and is then letting them escape multiple times, without sec, the ai, etc. noticing, and dealing with them (after the first 1-2 "attack waves" youd probably be on red already, armed and armored) then I feel like thats just a skill issue on secs part rather then the weapon being too powerfull, furthermore this could also be achieved with any of the not too expensive syndie weapons (ranging from a viper to a python) combined with an emagged lathe to supply ammo, where they can heal and reload inbetween strikes. If you are letting them get away multiple times so they can prepare for the next run, thats moreso on you then on the weapon. Keep in mind if we nerf the recharge to the degree it would take to fix the problem you are listing, itll be so long that you end up being better off running to a recharger rather then waiting it out. The whole point of the 10 second cooldown is to force you to disengage from the fight before you start recharging.

**Comment by SoulFN** [comment]
I think, wide needs 6 shots and narrow 8, because wide shot have more lasers and damage.

**Comment by BramvanZijp** [comment]
> I think, wide needs 6 shots and narrow 8, because wide shot have more lasers and damage.

It has equal damage, just more lasers aka more total damage on paper, but in practice the majority of those more lasers will miss in anything but point-blank range, which still leaves the narrow spread better for pretty much everything.

**Comment by BramvanZijp** [comment]
Made a very slight change & updated the branch.

**Comment by BramvanZijp** [comment]
2 Changes have been made:

Changing between the fire-modes will now properly update the HUD to reflect the changes in how many shots & max shots the weapon has, instead of you needing to unequip and reequip the item for those changes to take effect visually.

Also fixed the issue with self-recharging weapons sometimes failing to recharge their final shot completely, instead of applying a duct-tape fix to only the energy shotgun, no CL because this issue did not matter without the self recharge cooldown system.

**Comment by BramvanZijp** [comment]
Cleaned up the code regarding the ammo UI updating when switching firemode.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs [hunk 1]
```diff
@@ -9,5 +9,11 @@ public sealed partial class BatterySelfRechargerComponent : Component
         [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRecharge")] public bool AutoRecharge { get; set; }
 
         [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargeRate")] public float AutoRechargeRate { get; set; }
+
+        [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargePause")] public bool AutoRechargePause { get; set; } = false;
+
+        [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargePauseTime")] public float AutoRechargePauseTime { get; set; } = 0f;
+
+        [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargeCooldown")] public float AutoRechargeCooldown { get; set; } = 0f;
```
Comment (ScarKy0) [inline Content.Server/Power/Components/BatterySelfRechargerComponent.cs:17]: I recall being told that ViewVariables is no longer needed as DataField has that functionality. But I might be wrong.

**Review by ScarKy0 — APPROVED** [review]
Tested in game with no issues other than ones caused by battery not being predicted.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs (hunk 1 above)
Comment (BramvanZijp) [reply Content.Server/Power/Components/BatterySelfRechargerComponent.cs:17]: Copying what was done in that component originally to be safe.

**Comment by slarticodefast** [comment]
Might need to be adjusted to #30627

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs (hunk 1 above)
Comment (slarticodefast) [inline Content.Server/Power/Components/BatterySelfRechargerComponent.cs:17]:
```suggestion
        [DataField]
        public bool AutoRecharge { get; set; }

        [DataField]
        public float AutoRechargeRate { get; set; }

        [DataField]
        public bool AutoRechargePause { get; set; } = false;

        [DataField]
        public float AutoRechargePauseTime { get; set; } = 0f;

        [DataField]
        public float AutoRechargeCooldown { get; set; } = 0f;
```
Yeah, `ViewVariables(VVAccess.ReadWrite)` is included in DataField. The names are automatically converted to camelCase as well. And give the datafields summary strings for documentation what they do.
(I did not check the rest of the code)

**Comment by BramvanZijp** [comment]
> Might need to be adjusted to #30627

Solved.

**Comment by BramvanZijp** [comment]
Added a CL entry for the firemode HUD fix since I feel like forks should know what suddenly changed this (For example frontier and delta-v with their energy guns, x-01 multiphase guns, etc.)

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs [hunk 2]
```diff
@@ -6,8 +6,19 @@ namespace Content.Server.Power.Components
...
+        [DataField]
+        public bool AutoRechargePause { get; set; } = false;
+
+        [DataField]
+        public float AutoRechargePauseTime { get; set; } = 0f;
+
+        [DataField]
+        public float AutoRechargeCooldown { get; set; } = 0f;
     }
 }
```
Comment (metalgearsloth) [inline Content.Server/Power/Components/BatterySelfRechargerComponent.cs:24]: None of these are documented also pretty sure you don't even need most of these?

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 3]
```diff
@@ -145,6 +163,15 @@ public void SetCharge(EntityUid uid, float value, BatteryComponent? battery = nu
...
         }
 
+        public void SetChargeCooldown(EntityUid uid, float value, BatterySelfRechargerComponent? batteryself = null)
+        {
+            if (!Resolve(uid, ref batteryself))
+                return;
+
+            var old = batteryself.AutoRechargeCooldown;
+            batteryself.AutoRechargeCooldown = Math.Clamp(value, 0, 9999);
+        }
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:173]: Please don't do magic clamps just set the value, also this should really be in shared.

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 4]
```diff
@@ -110,6 +121,13 @@ public float UseCharge(EntityUid uid, float value, BatteryComponent? battery = n
             var newValue = Math.Clamp(0, battery.CurrentCharge - value, battery.MaxCharge);
             var delta = newValue - battery.CurrentCharge;
             battery.CurrentCharge = newValue;
+
+            if (TryComp<BatterySelfRechargerComponent>(uid, out var batterySelfRechargerComponent)){
+                if (batterySelfRechargerComponent.AutoRechargePause && batterySelfRechargerComponent.AutoRechargePauseTime > 0 && value > 0) {
+                    SetChargeCooldown(uid, Math.Clamp(batterySelfRechargerComponent.AutoRechargeCooldown, batterySelfRechargerComponent.AutoRechargePauseTime, 9999), batterySelfRechargerComponent);
+                }
+            }
+
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:130]: This formatting's really not great you're also double-clamping the value, once here and once in setchargecooldown. Value is also already checked 4 lines above so is pointless here.

Also please document it.

File: Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs [hunk 5]
```diff
@@ -99,13 +100,20 @@ private void SetFireMode(EntityUid uid, BatteryWeaponFireModesComponent componen
...
-            projectileBatteryAmmoProvider.FireCost = fireMode.FireCost;
+            var OldFireCost = projectileBatteryAmmoProviderComponent.FireCost;
+            projectileBatteryAmmoProviderComponent.Prototype = fireMode.Prototype;
+            projectileBatteryAmmoProviderComponent.FireCost = fireMode.FireCost;
+            float FireCostDiff = (float)fireMode.FireCost / (float)OldFireCost;
+            projectileBatteryAmmoProviderComponent.Shots = (int)Math.Round(projectileBatteryAmmoProviderComponent.Shots/FireCostDiff);
+            projectileBatteryAmmoProviderComponent.Capacity = (int)Math.Round(projectileBatteryAmmoProviderComponent.Capacity/FireCostDiff);
+            Dirty(uid, projectileBatteryAmmoProviderComponent);
+            var updateClientAmmoEvent = new UpdateClientAmmoEvent();
+            RaiseLocalEvent(uid, ref updateClientAmmoEvent);
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs:116]: This event is pointless just call updateammocount directly?

Also wouldn't it be better to just recalculate the shots from the batterycomponent directly? I feel like this is going to have rounding problems.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Power/EntitySystems/BatterySystem.cs (hunk 3 above)
Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs:173]: You are right, it should be in shared, but power is not in shared.

File: Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs (hunk 5 above)
Comment (BramvanZijp) [reply Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs:116]: The batterycomponent is not in shared, thats whats causing this entire issue.

Comment (BramvanZijp) [reply Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs]: Also, updateammocount only exists in client, so Im relatively sure calling it directly will cause more issues.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs (hunk 2 above)
Comment (BramvanZijp) [reply Content.Server/Power/Components/BatterySelfRechargerComponent.cs]: Fixed

File: Content.Server/Power/EntitySystems/BatterySystem.cs (hunk 3 above)
Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs]: Clamps removed, but I am not paid enough to move power to shared.

File: Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs (hunk 5 above)
Comment (BramvanZijp) [reply Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs]: Cannot apply changes due to the client/server barrier.

File: Content.Server/Power/EntitySystems/BatterySystem.cs (hunk 4 above)
Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs]: Fixed,

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 6]
```diff
@@ -100,6 +111,11 @@ private void OnEmpPulse(EntityUid uid, BatteryComponent component, ref EmpPulseE
         {
             args.Affected = true;
             UseCharge(uid, args.EnergyConsumption, component);
+            // Apply a cooldown to the entity's self recharge if needed to avoid it immediately self recharging after an EMP.
+            if (TryComp<BatterySelfRechargerComponent>(uid, out var batterySelfRechargerComponent))
+                if (batterySelfRechargerComponent.AutoRechargePause && batterySelfRechargerComponent.AutoRechargePauseTime > 0)
+                    if (args.EnergyConsumption > 0 && batterySelfRechargerComponent.AutoRechargePauseTime > batterySelfRechargerComponent.AutoRechargeCooldown)
+                        SetChargeCooldown(uid, batterySelfRechargerComponent.AutoRechargePauseTime, batterySelfRechargerComponent);
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:118]: This shouldn't be tightly coupled into other systems.

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 7]
```diff
@@ -110,6 +126,13 @@ public float UseCharge(EntityUid uid, float value, BatteryComponent? battery = n
             var newValue = Math.Clamp(0, battery.CurrentCharge - value, battery.MaxCharge);
             var delta = newValue - battery.CurrentCharge;
             battery.CurrentCharge = newValue;
+
+            // Apply a cooldown to the entity's self recharge if needed.
+            if (TryComp<BatterySelfRechargerComponent>(uid, out var batterySelfRechargerComponent))
+                if (batterySelfRechargerComponent.AutoRechargePause && batterySelfRechargerComponent.AutoRechargePauseTime > 0)
+                    if (value > 0 && batterySelfRechargerComponent.AutoRechargePauseTime > batterySelfRechargerComponent.AutoRechargeCooldown)
+                        SetChargeCooldown(uid, batterySelfRechargerComponent.AutoRechargePauseTime, batterySelfRechargerComponent);
+
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:135]: Same here.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs [hunk 8]
```diff
@@ -6,8 +6,29 @@ namespace Content.Server.Power.Components
...
+        /// <summary>
+        /// Should this entity stop automatically recharging if a charge is used?
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargePause")] public bool AutoRechargePause { get; set; } = false;
+
+        /// <summary>
+        /// How long should the entity stop automatically recharging if a charge is used?
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)] [DataField("autoRechargePauseTime")] public float AutoRechargePauseTime { get; set; } = 0f;
+
```
Comment (metalgearsloth) [inline Content.Server/Power/Components/BatterySelfRechargerComponent.cs:28]: 1. Would you ever add the component without autorecharge? Setting it to false as default also doesn't make sense.
2. Having both autorechargecooldown and autorechargepausetime overcomplicates it, better to just have one.
3. Please don't use datafield strings or properties unless necessary.

Comment (BramvanZijp) [reply Content.Server/Power/Components/BatterySelfRechargerComponent.cs:28]: 1. autoRecharge is a field that existed from before this PR.
2. Cooldown is the CURRENT cooldown, so ill remove the datafield string from it.
3. Tied into 2.

File: Content.Server/Power/EntitySystems/BatterySystem.cs (hunk 6 above)
Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs:118]: Clarify?

Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs]: Cleaned it up.

File: Content.Server/Power/EntitySystems/BatterySystem.cs (hunk 7 above)
Comment (BramvanZijp) [reply Content.Server/Power/EntitySystems/BatterySystem.cs]: Also cleaned.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs (hunk 8 above)
Comment (BramvanZijp) [reply Content.Server/Power/Components/BatterySelfRechargerComponent.cs:30]: Reworked it to use a timespan.

**Comment by BramvanZijp** [comment]
I am still able & willing to work on this PR, just to clarify.

File: Content.Server/Power/Components/BatterySelfRechargerComponent.cs (hunk 8 above)
Comment (metalgearsloth) [reply Content.Server/Power/Components/BatterySelfRechargerComponent.cs:30]: I mean autorechargepaused.

Also datafields still needed cleaning up.

**Review by metalgearsloth — APPROVED** [review]

**Review by Errant-4 — APPROVED** [review]
The design sounds reasonable and it probably does need a buff for a unique item

**Pull request merged**